vnc/vnc.c

```diff
@@ -586,6 +992,447 @@ make_color(int r, int g, int b, int bpp)
+
+    free_stream(s);
+
+    return error;
+}
+
+/**************************************************************************//**
+ * Tests if extended desktop size rect is an initial geometry specification
+ *
+ * This should be x == 0, but the specification says to treat undefined
+ * values as 0 also */
+static int
+rect_is_initial_geometry(int x, int y, int cx, int cy)
+{
+    return (x != 1 && x != 2);
+}
+
+/**************************************************************************//**
+ * Tests if extended desktop size rect is a reply to a request from us
+ */
+static int
+rect_is_reply_to_us(int x, int y, int cx, int cy)
+{
+    return (x == 1);
+}
+
+/**************************************************************************//**
+ * Handles the first framebuffer update from the server
+ *
+ * This is used to determine if the server supports resizes from
+ * us. See The RFB community wiki for details.
+ *
+ * If the server does support resizing, we send our client geometry over.
+ *
+ * @param v VNC object
+ * @return != 0 for error
+ */
+static int
+lib_framebuffer_first_update(struct vnc *v)
+{
+    int error;
+    struct vnc_screen_layout layout = {0};
+
+    error = find_matching_extended_rect(v,
+                                        rect_is_initial_geometry,
+                                        NULL,
+                                        NULL,
+                                        &layout);
+    if (error == 0)
+    {
+        if (layout.s != NULL)
+        {
+            log_message(LOG_LEVEL_DEBUG, "VNC server supports resizing");
+
+            /* Force the client geometry over to the server */
+            log_screen_layout(LOG_LEVEL_DEBUG, "Xvnc", &layout);
+
+            /*
+             * If we've only got one screen, and the other side has
+             * only got one screen, we will preserve their screen ID
+             * and any flags.  This may prevent us sending an unwanted
+             * SetDesktopSize message if the screen dimensions are
+             * a match. We can't do this with more than one screen,
+             * as we have no way to map different IDs
+             */
+            if (layout.count == 1 && v->client_layout.count == 1)
+            {
+                log_message(LOG_LEVEL_DEBUG, "VNC "
+                            "setting screen id to %d from server",
+                            layout.s[0].id);
+
+                v->client_layout.s[0].id = layout.s[0].id;
+                v->client_layout.s[0].flags = layout.s[0].flags;
+            }
+
+            if (vnc_screen_layouts_equal(&layout, &v->client_layout))
+            {
+                log_message(LOG_LEVEL_DEBUG, "Server layout is the same "
+                            "as the client layout");
+                v->initial_resize_status = VRS_DONE;
+            }
+            else
+            {
+                log_message(LOG_LEVEL_DEBUG, "Sending SetDesktopSize "
+                            "to change server layout");
```

Comment:
This could cause redundant  logs. Just below `send_set_desktop_size()` writes almost the same log.

```
[20200601-12:58:03] [DEBUG] Sending SetDesktopSize to change server layout
[20200601-12:58:03] [DEBUG] VNC Sending SetDesktopSize
```

For consistency with the opposite condition, I'd like to have the message like "Server layout differs from the client layout, changing server layout"  here.

Reply:
Also, isn't here the place to log OldLayout (current Xvnc layout)?

Reply:
Agree with both - OldLayout is now LOG_LEVEL_INFO